[Curl] Add support for HTTP alternative services

kshukuwa opened the pull request:
#### 2162ae8bd6047aced442b0d00edce9b270dec870
<pre>
[Curl] Add support for HTTP alternative services
<a href="https://bugs.webkit.org/show_bug.cgi?id=255017">https://bugs.webkit.org/show_bug.cgi?id=255017</a>

Reviewed by Alex Christensen.

We add support for the Alt-Svc feature to support HTTP/3
on the curl port in the future.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/curl/CurlContext.cpp:
(WebCore::CurlContext::clearAlternativeServicesStorageFile):
(WebCore::CurlContext::isAltSvcEnabled const):
(WebCore::CurlContext::isHttp2Enabled const):
(WebCore::CurlHandle::CurlHandle):
(WebCore::CurlHandle::enableAltSvc):
* Source/WebCore/platform/network/curl/CurlContext.h:
(WebCore::CurlContext::alternativeServicesStorageFile const):
(WebCore::CurlContext::setAlternativeServicesStorageFile):
* Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
(WebCore::alternativeServicesStorageFile):
(WebCore::NetworkStorageSession::NetworkStorageSession):
(WebCore::NetworkStorageSession::clearAlternativeServices):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::addStorageSession):
(WebKit::NetworkProcess::addWebsiteDataStore):
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteDataForOrigins):
(WebKit::NetworkProcess::deleteAndRestrictWebsiteDataForRegistrableDomains):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::NetworkSessionCocoa::hostNamesWithAlternativeServices const):
(WebKit::NetworkSessionCocoa::deleteAlternativeServicesForHostNames):
(WebKit::NetworkSessionCocoa::clearAlternativeServices):
* Source/WebKit/NetworkProcess/curl/NetworkSessionCurl.cpp:
(WebKit::NetworkSessionCurl::clearAlternativeServices):
* Source/WebKit/NetworkProcess/curl/NetworkSessionCurl.h:
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.cpp:
(WebKit::WebsiteDataRecord::addAlternativeServicesHostname):
* Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
* Source/WebKitLegacy/WebCoreSupport/NetworkStorageSessionMap.cpp:
(NetworkStorageSessionMap::ensureSession):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/263214@main">https://commits.webkit.org/263214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f29e5ee2ecef8c5d5ad44fa7147aad95480c6852

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4067 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5413 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/4224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3950 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4155 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4042 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4151 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4022 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4207 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3582 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5316 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1709 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3558 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5645 "1 flakes 146 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3277 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3533 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3618 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5044 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/3750 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4024 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4054 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3541 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3558 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1003 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/965 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3591 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/4147 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3821 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1115 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/UIProcess/API/C/curl/WKWebsiteDataStoreConfigurationRefCurl.h:
```diff
@@ -0,0 +1,40 @@
+#endif
+
+WK_EXPORT WKStringRef WKWebsiteDataStoreConfigurationCopyAlternativeServicesStorageFile(WKWebsiteDataStoreConfigurationRef configuration);
+WK_EXPORT void WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile(WKWebsiteDataStoreConfigurationRef configuration, WKStringRef cacheFile);
```
Do applications have to use this API to enable Alt-Svc?
Do you really need to set a filename?
What about an idea of using an existing API `WKContextConfigurationSetDiskCacheDirectory` and a fixed file name "altsvc-cache.txt"?

kshukuwa replied:
> Do applications have to use this API to enable Alt-Svc?

Some applications may not require Alt-Svc. Therefore I would like to be able to enable/disable this feature.

> Do you really need to set a filename?
> What about an idea of using an existing API WKContextConfigurationSetDiskCacheDirectory and a fixed file name "altsvc-cache.txt"?

Disk Cache and Alt-Svc are completely unrelated. I think this is a very confusing specification for users.

kshukuwa replied:
Would you like to set a default filename?

For example:
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h#L371
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp#L2038-L2046
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp#L92

kshukuwa replied:
I've tried creating a patch that sets the default filename, and I think this is the simplest way to do it.
How do you think ?

```diff
diff --git a/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h b/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h
index 0e46c4e936d2..29aa986a8640 100644
--- a/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h
+++ b/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h
@@ -311,6 +311,7 @@ public:
 
 #if USE(CURL)
     const String& resolvedAlternativeServicesStorageFile() const { return m_resolvedConfiguration->alternativeServicesStorageFile(); }
+    static String defaultAlternativeServicesStorageFile(const String& baseCacheDirectory = nullString());
 
     void setNetworkProxySettings(WebCore::CurlProxySettings&&);
     const WebCore::CurlProxySettings& networkProxySettings() const { return m_proxySettings; }
diff --git a/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp b/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp
index 5f4b9419efd1..eafe00664d38 100644
--- a/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp
+++ b/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp
@@ -110,6 +110,9 @@ void WebsiteDataStoreConfiguration::initializePaths()
     setCookieStorageFile(WebsiteDataStore::defaultCookieStorageFile(m_baseDataDirectory));
     setSearchFieldHistoryDirectory(WebsiteDataStore::defaultSearchFieldHistoryDirectory(m_baseDataDirectory));
 #endif
+#if USE(CURL)
+    setAlternativeServicesStorageFile(WebsiteDataStore::defaultAlternativeServicesStorageFile(m_baseDataDirectory));
+#endif
 }
 
 Ref<WebsiteDataStoreConfiguration> WebsiteDataStoreConfiguration::copy() const
diff --git a/Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp b/Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp
index 0bde68197912..2cd39279d88a 100644
--- a/Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp
+++ b/Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp
@@ -40,6 +40,15 @@ void WebsiteDataStore::platformSetNetworkParameters(WebsiteDataStoreParameters&
     parameters.networkSessionParameters.proxySettings = m_proxySettings;
 }
 
+String WebsiteDataStore::defaultAlternativeServicesStorageFile(const String& baseCacheDirectory)
+{
+#if PLATFORM(PLAYSTATION)
+    return nullString();
+#else
+    return cacheDirectoryFileSystemRepresentation("altsvc-cache.txt"_s, baseCacheDirectory);
+#endif
+}
+
 void WebsiteDataStore::setNetworkProxySettings(WebCore::CurlProxySettings&& proxySettings)
 {
     m_proxySettings = WTFMove(proxySettings);
```

kshukuwa replied:
Fixed to set default filename.

fujii replied:
I think Alt-Svc cache is a disk cache of network response.
How other browsers do? Do they have a UI to clear Alt-Svc cache? How to clear Alt-Svc cache?
How to enable/disable Alt-Svc with the new API? Call WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile with empty string to disable? It's disabled by default only for PlayStation port?
How to clear Alt-Svc cache? Remove the file by using a system API?
Why do you need enable/disable Alt-Svc? Do you need API to enable/disable other HTTP features?

kshukuwa replied:
> How other browsers do?

The Cocoa port is designed to specify the Alt-Svc storage directory.
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreConfiguration.h#L97

> Do they have a UI to clear Alt-Svc cache?

Delete it as part of the WebsiteData.

> How to clear Alt-Svc cache?
> Remove the file by using a system API?

For Cocoa port, it will be removed using the removeDataOfTypes function.
https://github.com/WebKit/WebKit/blob/main/Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm#L706

In the case of Curl port, it is necessary to prepare a dedicated API like PR below.
https://github.com/WebKit/WebKit/pull/6200

However, as Alex mentioned in this PR, it would be better to have an API like removeDataOfTypes rather than writing a dedicated API. As such, I did not provide an API for removal in this PR.

kshukuwa replied:
> Call WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile with empty string to disable?

Hmm.. If WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile is an empty string, it will not be possible to overwrite its value. This means that once the default value is set it cannot be turned off.

 https://github.com/WebKit/WebKit/pull/12404/files#diff-f5dc1cba4ced1243c9430dea1be6d75874b4b8d452fd989cf498d3810c164198R382

kshukuwa replied:
> However, as Alex mentioned in this PR, it would be better to have an API like removeDataOfTypes
> rather than writing a dedicated API. As such, I did not provide an API for removal in this PR.

Since such APIs have not been created yet, I have prepared a dedicated API `WKWebsiteDataStoreRemoveAlternativeServicesStorageFile`.

kshukuwa replied:
Removed WKWebsiteDataStoreConfigurationSetAlternativeServicesStorageFile.

fujii commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -310,6 +310,9 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
 
 #if USE(CURL)
+    const String& resolvedAlternativeServicesStorageFile() const { return m_resolvedConfiguration->alternativeServicesStorageFile(); }
+    static String defaultAlternativeServicesStorageFile(const String& baseCacheDirectory = nullString());
```
Remove `= nullString()`. This is not used.

kshukuwa replied:
The implementation is aligned with other defaultXXX functions.

https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h#L365-L386

kshukuwa replied:
Fixed

achristensen07 commented on Source/WebKit/UIProcess/API/C/curl/WKWebsiteDataStoreRefCurl.h:
```diff
@@ -36,6 +36,9 @@ WK_EXPORT void WKWebsiteDataStoreEnableDefaultNetworkProxySettings(WKWebsiteData
 WK_EXPORT void WKWebsiteDataStoreEnableCustomNetworkProxySettings(WKWebsiteDataStoreRef, WKURLRef proxyUrl, WKStringRef ignoreHosts);
 WK_EXPORT void WKWebsiteDataStoreDisableNetworkProxySettings(WKWebsiteDataStoreRef);
 
+typedef void (*WKWebsiteDataStoreRemoveAlternativeServicesStorageFileCallback)(void* functionContext);
```
It would be better to be heading towards an API design closer to WKWebsiteDataStore.allWebsiteDataTypes and removeDataOfTypes.

achristensen07 commented on Source/WebKit/UIProcess/API/C/curl/WKWebsiteDataStoreConfigurationRefCurl.cpp:
```diff
@@ -0,0 +1,40 @@
+#include "WKAPICast.h"
+#include "WebsiteDataStoreConfiguration.h"
+
+WKStringRef WKWebsiteDataStoreConfigurationCopyAlternativeServicesStorageFile(WKWebsiteDataStoreConfigurationRef configuration)
```
Rather than introducing new API to set new directories on WKWebsiteDataStoreConfiguration, I think it would be good to be working towards a design like WKWebsiteDataStore.dataStoreForIdentifier where you make a data store with a directory with a UUID name and put all your storage there.  That way, if an existing application doesn't update to use the new API, it will still manage all its storage types correctly.

kshukuwa replied:
I will check it.

kshukuwa replied:
It looks like the dataStoreForIdentifier function creates a UUID directory under NSLibrarydirectory. However, on other platforms, users may want to change the directory location, which corresponds to the NSLibraryDirectory.

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp#L262

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp#L62-L63

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm#L312-L315

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm#L252

kshukuwa replied:
Would it be okay to add the WKWebsiteDataStoreConfigurationCreateWithBaseDirectory function to WKWebsiteDataStoreConfiurationRef, as it would be convenient to be able to specify the base directory on the other ports?

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h#L55-L59

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.cpp#L80-L81

https://github.com/WebKit/WebKit/blob/0646de88e5b9e9e657741c81b75e51e353ff35ae/Source/WebKit/UIProcess/API/C/WKWebsiteDataStoreConfigurationRef.h#L37

kshukuwa replied:
Then, we save the altsvc-chache.txt file under the m_baseCacheDirectory directory.

achristensen07 commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -145,6 +145,10 @@ class NetworkStorageSession : public CanMakeWeakPtr<NetworkStorageSession> {
     WEBCORE_EXPORT void setCookieAcceptPolicy(CookieAcceptPolicy) const;
     WEBCORE_EXPORT void setProxySettings(const CurlProxySettings&);
+
+    WEBCORE_EXPORT void setAlternativeServicesStorageFile(const String& cacheFile) const;
```
Can we put this in the constructor instead?

kshukuwa replied:
Yes. I will fix it.

kshukuwa replied:
Fixed

achristensen07 commented on Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
```diff
@@ -232,6 +232,16 @@ void NetworkStorageSession::setProxySettings(const CurlProxySettings& proxySetti
 
+void NetworkStorageSession::setAlternativeServicesStorageFile(const String& cacheFile) const
+{
+    CurlContext::singleton().setAlternativeServicesStorageFile(cacheFile);
```
I think you should be working towards making CurlContext no longer a singleton.  That's what NetworkStorageSession is for.

kshukuwa replied:
I will fix it in a separate PR.

achristensen07 commented on Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
```diff
@@ -75,7 +75,7 @@ WebsiteDataProcessType WebsiteData::ownerProcess(WebsiteDataType dataType)
     case WebsiteDataType::PrivateClickMeasurements:
         return WebsiteDataProcessType::Network;
-#if HAVE(CFNETWORK_ALTERNATIVE_SERVICE)
+#if HAVE(CFNETWORK_ALTERNATIVE_SERVICE) || USE(CURL)
```
Maybe we should introduce a HAVE(ALTERNATIVE_SERVICE) macro that is true if we have CFNetwork alt-svc support or cURL.  Fine to do in a separate PR.

kshukuwa replied:
~~OK, I will fix it in a separate PR.~~

kshukuwa replied:
Fixed in this PR.

achristensen07 commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -378,6 +378,12 @@ void WebsiteDataStore::resolveDirectoriesIfNecessary()
         m_resolvedConfiguration->setCookieStorageFile(FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(m_configuration->cookieStorageFile())));
     }
+#if USE(CURL)
+    if (!m_configuration->alternativeServicesStorageFile().isEmpty()) {
```
We already have this above:
if (!m_configuration->alternativeServicesDirectory().isEmpty())
Can't we just use that?

kshukuwa replied:
I will fix it.

kshukuwa replied:
Fixed

achristensen07 commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreConfiguration.h:
```diff
@@ -318,6 +323,9 @@ class WebsiteDataStoreConfiguration : public API::ObjectImpl<API::Object::Type::
     RetainPtr<CFDictionaryRef> m_proxyConfiguration;
 #endif
+#if USE(CURL)
+    String m_alternativeServicesStorageFile;
```
This is redundant with m_alternativeServicesDirectory

kshukuwa replied:
I will fix it.

kshukuwa replied:
Fixed

achristensen07 commented on Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
```diff
@@ -77,11 +84,15 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
     , m_cookieDatabase(makeUniqueRef<CookieJarDB>(sessionID.isEphemeral() ? ":memory:"_s : defaultCookieJarPath()))
 {
+    if (!alternativeServicesDirectory.isEmpty()) {
+        FileSystem::makeAllDirectories(alternativeServicesDirectory);
```
I think you'll find this is unnecessary because it's done in the UI process before telling the network process to make this NetworkStorageSession.  It's in cURL-specific code, though, so I don't have strong opinions.

kshukuwa replied:
I tried it, but the directory was not created. Since SandboxExtension is not enabled for non-Cocoa, it seems that the directory is not created. So, I leave this code as-is.

kshukuwa committed:
[Curl] Add support for HTTP alternative services
https://bugs.webkit.org/show_bug.cgi?id=255017

Reviewed by Alex Christensen.

We add support for the Alt-Svc feature to support HTTP/3
on the curl port in the future.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/curl/CurlContext.cpp:
(WebCore::CurlContext::clearAlternativeServicesStorageFile):
(WebCore::CurlContext::isAltSvcEnabled const):
(WebCore::CurlContext::isHttp2Enabled const):
(WebCore::CurlHandle::CurlHandle):
(WebCore::CurlHandle::enableAltSvc):
* Source/WebCore/platform/network/curl/CurlContext.h:
(WebCore::CurlContext::alternativeServicesStorageFile const):
(WebCore::CurlContext::setAlternativeServicesStorageFile):
* Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp:
(WebCore::alternativeServicesStorageFile):
(WebCore::NetworkStorageSession::NetworkStorageSession):
(WebCore::NetworkStorageSession::clearAlternativeServices):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::addStorageSession):
(WebKit::NetworkProcess::addWebsiteDataStore):
(WebKit::NetworkProcess::fetchWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteData):
(WebKit::NetworkProcess::deleteWebsiteDataForOrigins):
(WebKit::NetworkProcess::deleteAndRestrictWebsiteDataForRegistrableDomains):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.cpp:
(WebKit::NetworkSessionCreationParameters::encode const):
(WebKit::NetworkSessionCreationParameters::decode):
* Source/WebKit/NetworkProcess/NetworkSessionCreationParameters.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::NetworkSessionCocoa):
(WebKit::NetworkSessionCocoa::hostNamesWithAlternativeServices const):
(WebKit::NetworkSessionCocoa::deleteAlternativeServicesForHostNames):
(WebKit::NetworkSessionCocoa::clearAlternativeServices):
* Source/WebKit/NetworkProcess/curl/NetworkSessionCurl.cpp:
(WebKit::NetworkSessionCurl::clearAlternativeServices):
* Source/WebKit/NetworkProcess/curl/NetworkSessionCurl.h:
* Source/WebKit/Shared/WebsiteData/WebsiteData.cpp:
(WebKit::WebsiteData::ownerProcess):
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.h:
* Source/WebKit/Shared/WebsiteData/WebsiteDataType.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataRecordInternal.h:
(WebKit::toWebsiteDataType):
(WebKit::toWKWebsiteDataTypes):
* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataRecord.cpp:
(WebKit::WebsiteDataRecord::addAlternativeServicesHostname):
* Source/WebKit/UIProcess/WebsiteData/curl/WebsiteDataStoreCurl.cpp:
(WebKit::WebsiteDataStore::platformSetNetworkParameters):
* Source/WebKitLegacy/WebCoreSupport/NetworkStorageSessionMap.cpp:
(NetworkStorageSessionMap::ensureSession):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsiteDataStoreCustomPaths.mm:
(TEST):

Canonical link: https://commits.webkit.org/263214@main
